main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala

```diff
@@ -77,6 +77,11 @@ class LanguageServer(port: Int) extends WebSocketServer(new InetSocketAddress("l
     */
   val sources: mutable.Map[String, String] = mutable.Map.empty
 
+  /**
+    * Sources which have been removed since the last check
+    */
+  val removedSources: mutable.Map[String, String] = mutable.Map.empty
```

Comment:
What's the reason we need to maintain this state?

Also, what happens if a file is deleted and re-added? Presumably we need to remove it from this map then?

Reply:
It was by analogy to the way that source files are added. When a source file is added by VSCode, the language server just adds it to the `sources` map. It's only added to the Flix compiler (via `addSourceCode`) when `procesCheck` is called.

I assumed that there was a good reason for it being done this way (possibly concurrency?).

But now I think about it, I can 't see why it is. The language server only handles a single message a time (this has to be true, otherwise using unprotected mutable maps would be unsafe).

So we can get rid of `removeSources`, and also get rid of the code that adds sources to the Flix compiler in `processCheck`, simply adding and removing them directly within the `Request.AddUri` and `Request.RemUri` cases.

I'll make that change (unless there's something I'm missing)?

Reply:
(oh, regarding the question of it being removed and re-added, that doesn't need special handling because we do the removes before adds within `processCheck`, but if I make the change suggested above, then it's moot and much simpler).

Reply:
> It was by analogy to the way that source files are added. When a source file is added by VSCode, the language server just adds it to the `sources` map. It's only added to the Flix compiler (via `addSourceCode`) when `procesCheck` is called.
> 
> I assumed that there was a good reason for it being done this way (possibly concurrency?).
> 
> But now I think about it, I can 't see why it is. The language server only handles a single message a time (this has to be true, otherwise using unprotected mutable maps would be unsafe).
> 
> So we can get rid of `removeSources`, and also get rid of the code that adds sources to the Flix compiler in `processCheck`, simply adding and removing them directly within the `Request.AddUri` and `Request.RemUri` cases.
> 
> I'll make that change (unless there's something I'm missing)?

I think that should be possible.